src/passes/OptimizeInstructions.cpp

```diff
@@ -1738,16 +1743,99 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // (x op C1) + (x op C2)   ==>   x * (C1' + C2')
+  // (x op C1) - (x op C2)   ==>   x * (C1' - C2')
+  // where op = (`*`|`<<`)
+  // x * y + x * z   ==>   (y + z) * x
+  // x * y - x * z   ==>   (y - z) * x
+  Expression* optimizeBinaryDistributions(Binary* curr) {
+    using namespace Match;
+    assert(curr->type.isInteger());
+
+    if (curr->op == Abstract::getBinary(curr->type, Abstract::Add) ||
+        curr->op == Abstract::getBinary(curr->type, Abstract::Sub)) {
+      Expression* ll = nullptr;
+      Expression* rl = nullptr;
+      Expression* lr = nullptr;
+      Expression* rr = nullptr;
+      Binary* left = nullptr;
+      Binary* right = nullptr;
+
+      if (matches(curr->left, binary(&left, any(&ll), any(&lr))) &&
+          matches(curr->right, binary(&right, any(&rl), any(&rr))) &&
+          !effects(curr->right).hasSideEffects()) {
+        // canonicalize
+        // (x << C1) op (x << C2)
+        // to
+        // (x * (1 << C1)) op (x * (1 << C2))
+        bool eqLLRL = ExpressionAnalyzer::equal(ll, rl);
+        if (eqLLRL) {
+          Const *c1, *c2;
+          auto type = left->type;
+          if (matches(left, binary(Abstract::Shl, any(), constant(&c1)))) {
+            left->op = Abstract::getBinary(type, Abstract::Mul);
+            c1->value = Literal::makeFromInt32(1, type).shl(c1->value);
+          }
```

Comment:
Instead of nesting matches inside other matches, it would be better to have the top-level matches be as specific as possible. So here, the first top-level match should specifically check for `Abstract::Shl`.

Reply:
Yes, but this transform make sense only for this specific situation and otherwise will produce regression. Ofc I could move it out put before main transform and do full match for canonicalization and actual transform, but in my opinion it's double work which could be avoided. But I'm ok for this if it increase understanding

Reply:
I think repeating the full match makes the code easier to read, even though it does the work twice.

Reply:
ok